### DISCUSSION TITLE
Conditionally-Available Block Types

### DISCUSSION THREAD

**Discussion by politicoder** [payloadcms/payload#7064] [Q&A]
To my knowledge what I'm attempting to do is not currently possible in Payload, but I would love to be told otherwise.

My agency is pretty much the exact use-case described in the [Multi-Tenant Docs](https://payloadcms.com/blog/how-to-build-a-multi-tenant-app-with-payload); we currently host hundreds of custom WordPress sites that follow similar, but not identical, design patterns. For user-editable page layouts we use Advanced Custom Fields' "Flexible Content" field, which is functionally very similar to Payload's Blocks field.

I am sold on Payload and want very much to break free of WordPress on new sites going forward. Hosting each Payload site as a fully independent repo like we do with WordPress seems impractical from a server configuration standpoint; I would rather take advantage of Payload's multi-tenant capabilities and give each client a login to one giant Payload app as described in the above-linked docs.

The trouble is that while every site in the app has a Pages collection with a Blocks field for its layout, every client should NOT have access to every type of Block. There are a handful of universal blocks that every site will have, but in general, a big part of our design process is custom-designing blocks for individual websites. I need a way to designate any given Block as only being visible to certain Users. _How_ that's implemented, I don't really care; setting available blocks as a hasMany multiselect field on the Sites collection, doing it all in code, either way is fine. But if every tenant always has access to every block in the codebase, I'm not sure Payload is viable for our use case.

Any insight is appreciated, thanks!

**Comment by AshtarCodes** [comment]
I have a need for conditional blocks also. One of my blocks is a list of tabs that each accept blocks as content. If a user wants to use this, other blocks need to be unavailable at the top level, as all content should be under a tab in this pattern. 

My use case is a single pages collection that allows selecting from a list of templates, and conditionally modifying the templates. That includes using the template with tabs or without tabs. Having conditional blocks would clean up the implementation significantly.

**Comment by roybarber** [comment]
@politicoder  @AshtarCodes Did either of you find a solution?

**Comment by vonkoff** [comment]
I'm trying to find an answer on this as well. Hopefully a solution pops up soon.

**Comment by Collexi-Dev** [comment]
~_This discussion was linked in a [Discord thread](https://discord.com/channels/967097582721572934/1313579475614695474)._~

~I haven't set up multi-tenancy myself and I feel like I'm missing a piece of information here, but I do want to find out what since I'm most likely going to end up in the same scenario soon.~

~You want to show / hide blocks based on certain parameters?~

~If you make `blocks` a function:~
```
{
  name: 'layout',
  type: 'blocks',
  required: true,
  blocks: getConditionalBlocks(false),
},
```
  
~And have that return the correct blocks:~
```
function getConditionalBlocks(allowFormBlock: boolean): Block[] {
  const blocks = [MediaBlock, SpacerBlock];

  if (allowFormBlock) {
    blocks.push(BasicMessageContactFormBlock);
  }

  return blocks;
}
```

~If you replace the bool with a role check and define allowed blocks for specific roles, does this not achieve what you are looking for?~

Edit: Ok no, after experimenting with it I can see the problems, I have an idea for a solution though, testing it at the moment.

**Comment by Collexi-Dev** [comment]
So, after some experimenting I've come up with a possibility, it's currently not the cleanest and could use some finetuning as well as some more abstract / generic implementations because in the way I've set it up now it could quickly become messy, but nothing that can't be solved with some more code splitting.

This solution assumes there's a `tenant` value on `user` but this could be anything.

The main takeaway here is to have multiple `blocks` fields, since conditionally or dynamically modifying the `blocks` indeed doesn't seem possible, or I at least couldn't find how.

So, for example, in your pages collection you can add:
```ts
  {
    name: 'layout',
    type: 'blocks',
    blocks: AllBlocks,
    admin: {
      disabled: true,
    },
  },
  {
    name: 'collexiLayout',
    type: 'blocks',
    required: true,
    blocks: getBlocksForTenant('Collexi'),
    admin: {
      condition: (data, siblingData, { user }) => {
        return user.tenant === 'Collexi';
      },
    },
    hooks: {
      afterChange: [saveLayoutBlocksAfterChange],
    },
  },
  {
    name: 'filteredLayout',
    type: 'blocks',
    required: true,
    blocks: getFilteredBlocks(),
    admin: {
      condition: (data, siblingData, { user }) => {
        return user.tenant !== 'Collexi';
      },
    },
    hooks: {
      afterChange: [saveLayoutBlocksAfterChange],
    },
  },
```
The `layout` field is the one that will be used for rendering, while the other layouts are used in the CMS.
The `layout` field `blocks` should also include all of the available blocks for type generation.

The other layouts blocks should only contain the blocks available for that tenant, you can choose how you want to define these.
You could add separate functions and separately define blocks for tenants:
```ts
export function getCollexiBlocks() {
  return [BasicMessageContactFormBlock];
}

export function getFilteredBlocks() {
  return [MediaBlock, SpacerBlock];
}
```

Or you could extend `Block`:
```ts
export interface TenantBlock extends Block {
  allowedTenants: string[];
}
```
Use `TenantBlock` for your block config type and add the `allowedTenants`:
```ts
export const BasicMessageContactFormBlock: TenantBlock = {
  allowedTenants: ['Collexi'],
  ...
};
```
Then you could use:
```ts
export function getBlocksForTenant(tenant: string): Block[] {
  return AllBlocks.filter((block) => {
    return block.allowedTenants?.includes(tenant);
  });
}
```
And call it `blocks: getBlocksForTenant('Collexi')`.

The `saveLayoutBlocksAfterChange` field hook syncs the tenant `blocks` with the `blocks` used for rendering:
```ts
import {
  filterInvalidBlocks,
  getBlocksForTenant,
} from '../pageFields/getBlocks';
import { FieldHook } from 'payload';

export const saveLayoutBlocksAfterChange: FieldHook = async ({
  data,
  originalDoc,
  req,
}) => {
  if (!req.user || !req.user.tenant) {
    console.log(
      'Skipping filtering as no authenticated user with an assigned tenant is found.'
    );
    return data;
  }

  const tenant = req.user.tenant;
  console.log('Processing layout for tenant:', tenant);

  const allowedBlocks = getBlocksForTenant(tenant);

  if (tenant === 'Collexi') {
    data.builderContent.layout = filterInvalidBlocks(
      originalDoc.builderContent.collexiLayout,
      allowedBlocks
    );
  }

  if (tenant === 'Filtered') {
    data.builderContent.layout = filterInvalidBlocks(
      originalDoc.builderContent.filteredLayout,
      allowedBlocks
    );
  }

  return data;
};
```

I've added a `filterInvalidBlocks` here in case certain blocks would become unavailable for a tenant and they need to be removed, i.e. when you have "basic" and "premium" blocks and a tenant his premium subscription expires:
```ts
export function filterInvalidBlocks(blocks: any[], allowedBlocks: Block[]) {
  return blocks.filter((block) =>
    allowedBlocks.some((allowedBlock) => allowedBlock.slug === block.blockType)
  );
}
```

I also have a collection `beforeChange` hook that is similar to the field `afterChange` hook for extra validation and filtering. The field `afterChange` hook is required for continuous updating in case you want live previews to work.

And then I also added a collection `afterRead` hook which basically again, filters invalid blocks. The 3 hooks together is probably overkill and will be cleaned up, but this is currently just the result of experimenting and the end result of going down a small rabbit hole.

To improve maintainability for tens or hundreds of tenants you would probably want to add a helper function that iterates over all available tenants and build the `blocks` fields dynamically.

Does this offer a solution to what you were looking for?

**Comment by bhofstaetter** [comment]
I will soon be facing the same challenge. I have implemented something similar elsewhere, but with a SelectField. Depending on the current tenant, different options should be available.

My solution there was to overwrite the field component. I think that worked quite well.

My hope so far is that it will work the same way with the block field or the editor field (we only use the editor to make blocks available).

This is my code for the "tenant aware select field", maybe it can serve as an inspiration for your implementation.

templateField.ts

```typescript
const t = createTranslator('fields:templateField');

export const templateField = <Group extends TemplateGroupKey>(args: TemplateFieldArgs<Group>): TextField => {
    const {defaultValue, label, hidden, templateGroupKey, templateGroupEntryKey} = args;

    return {
        name: 'template',
        label: label ?? t('label'),
        type: 'text',
        hidden,
        required: true,
        defaultValue: defaultValue ?? 'default',
        admin: {
            components: {
                Field: {
                    path: '@/backend/fields/templateField/components/TenantAwareTemplateField',
                    serverProps: {
                        templateGroupKey,
                        templateGroupEntryKey,
                    },
                },
            },
        },
    };
};
```

TenantAwareTemplateField.tsx

```typescript
import {SelectField} from '@payloadcms/ui';
import {SelectFieldServerComponent} from 'payload';
import {findTenantByRequest} from '@/backend/api/local/findTenantByRequest';
import {getTenantAwareTemplateOptions} from '@/backend/utils/getTenantAwareTemplateOptions';

const TenantAwareTemplateField: SelectFieldServerComponent = async (props) => {
    const {schemaPath, clientField, req, templateGroupKey, templateGroupEntryKey} = props;

    const path = (props?.path || props?.field?.name || '') as string;

    const tenant = await findTenantByRequest({req});

    const options = getTenantAwareTemplateOptions({
        group: templateGroupKey as any,
        entry: templateGroupEntryKey as any,
        companyNumber: tenant?.companyNumber,
    });

    const field = {
        ...clientField,
        options,
    };

    return clientField.hidden || options.length <= 1 ? (
        <></>
    ) : (
        <SelectField field={field} path={path} schemaPath={schemaPath} />
    );
};

export default TenantAwareTemplateField;
```

**Comment by Bobeta** [comment]
The workaround we are doing right now is conditionally rendering the entire Editor and they have different blocks defined. Then we add a hook to copy the content of that editor to the "content" field in the API upon saving.

**Comment by kevmiota** [comment]
thanks a lot for the inspiration @bhofstaetter 

We're going with a wrapper component for the block field and injecting our own logic of available blocks based on a separate template field.
Maybe this will help someone:

BlocksFieldWrapper.tsk
```typescript
'use client'
import { useFormFields, BlocksField } from '@payloadcms/ui'
import React from 'react'

interface AllowedBlocks {
  [key: string]: string[]
}

interface BlocksFieldWrapperProps {
  allowedBlocks: AllowedBlocks
  field: any
  path: string // Required by BlocksField
  [key: string]: any // Additional props
}

export const BlocksFieldWrapper: React.FC<BlocksFieldWrapperProps> = ({ allowedBlocks, ...props }) => {
  const templateField = useFormFields(([fields]) => fields.template)

  let { field } = props

  const newField = {
    ...field,
    blocks: field.blocks.filter(({ slug }: { slug: string }) => {
      if (templateField && Array.isArray(allowedBlocks[templateField.value as string])) {
        return allowedBlocks[templateField.value as string].includes(slug)
      }
      return true
    }),
  }

  return <BlocksField {...props} field={newField} />
}

```

for example in page collection (Pages.ts) you can define which block slugs are allowed for the selected template field value

```typescript
{
  ...
  fields: [
    {
      name: 'blocks',
      type: 'blocks',
      blocks: [Text, Image, Table],
      admin: {
        components: {
          Field: {
            path: '@/components/BlockField/BlocksFieldWrapper#BlocksFieldWrapper',
            clientProps: {
              allowedBlocks: {
               home: ['text', 'image'],
               standard: ['text', 'table'],
              }
            }
          }
        }
      },
    },
 ]
}
```

This will actually hide the blocks depending on the template field. However, this will not delete already existing blocks. You'd have to create a separate beforeChange hook to filter out not allowed blocks before saving. This approach can be applied for a tenant field as well or even load a list of allowed blocks

**Reply by mjaysona** [reply]
This is currently how I implement conditional blocks but how do you update the value of your blocks field in your collection?

**Comment by ghost** [comment]
For anyone who stumbles across this like I did. Here's how I handled it, with nested blocks. Though it is a quick and dirty implementation. 

You can cut up path to use for the useFormFields path. I go back 2, but you can set it however deeply nested the block is. Then set the field you want the value for in text.

I also recommend setting up a filter function for each block that returns a boolean, then you can just pass a registry of them to the custom component and filter by slug. Makes it a lot easier to handle filtering.

### Component
```tsx
'use client'
import { useFormFields, BlocksField } from '@payloadcms/ui'
import React from 'react'
import { AllActionBlocksOptions } from '../blocks'
import { ActionBlockType, ActionBlockFilter } from '../blocks'

interface BlocksFieldWrapperProps {
  placement: 'content' | 'trigger'
  field: any
  path: string // Required by BlocksField
  [key: string]: any // Additional props
}

export const BlocksFieldWrapper: React.FC<BlocksFieldWrapperProps> = ({ ...props }) => {
  // Get the parent block path (remove last two segments)
  const pathParts = props.path.split('.')
  const parentBlockPath = pathParts.slice(0, -2).join('.')

  // Get the parent block's data
  const { value: blockType } = useFormFields(([fields]) => fields[`${parentBlockPath}.blockType`])
  const { value: type } = useFormFields(([fields]) => fields[`${parentBlockPath}.type`])
  const { value: buttonType } = useFormFields(([fields]) => fields[`${parentBlockPath}.buttonType`])
  const { value: portalType } = useFormFields(([fields]) => fields[`${parentBlockPath}.portalType`])
  const { value: formType } = useFormFields(([fields]) => fields[`${parentBlockPath}.formType`])
  const { value: inputType } = useFormFields(([fields]) => fields[`${parentBlockPath}.inputType`])

  const block = {
    placement: props?.placement,
    blockType,
    type,
    buttonType,
    portalType,
    formType,
    inputType,
  } as ActionBlockFilter

  let { field } = props

  const newField = {
    ...field,
    blocks: field.blocks.filter(({ slug }: { slug: ActionBlockType }) => {
      return AllActionBlocksOptions[slug].filter({ block })
    }),
  }

  return <BlocksField {...props} field={newField} />
}
```

### Field
```tsx
{
  type: 'blocks',
  name: 'actionBlocks',
  blocks: AllActionBlocks,
  admin: {
    initCollapsed: true,
    components: {
      Field: {
        path: '@/proLaico/actions/customComponents/actionBlock#BlocksFieldWrapper',
        clientProps: {
          placement: prefix,
        },
      },
    },
  },
},
```

### Example Filter Function
```tsx
filter: (params) => {
  const { block } = params
  return block.blockType === 'atomic' && block.type === 'button' && block.buttonType === 'regular'
}
```

**Comment by DanRibbens** [comment]
This was implemented in core, you can read the docs here: https://payloadcms.com/docs/fields/blocks#conditional-blocks

**Reply by Collexi-Dev** [reply]
Great, thanks for the follow up!

**Reply by mckenzie-boyd** [reply]
Will this also be added to Tab fields? Right now, I am running into a similar issue to conditionally render a tab in a group of tabs based on a feature flag that our server grabs. Considering I cannot use async/await in the admin -> condition, it has been a real headache. Maybe I am not understanding correctly how to implement this?

`  admin: {
    condition: () => await FeatureFlagEnabled,
  },
`
Will not work, as you cannot make it dynamic.
Is there a standard way of doing this that I am missing?

**Reply by truuman** [reply]
@mckenzie-mri a possible workaround to your problem is to have a `virtual` field `hidden` in `admin` with an `afterRead` hook that is populated with the FeatureFlag value you need for the conditional field

That way the condition function can be synchronous

**Reply by mckenzie-boyd** [reply]
I actually found a 'correct' way to do it! What you mentioned is what I tried to do but kept running into weird issues. My solution was as follows:

1. My Utils file was defaulted to 'client-side', my flag manager was 'server-side' (Calling it here would cause node fs errors). First step was to create a new Utils file with the 'use server' directive at the top, marking it for 'server-side' use only.

2. In my server utils file, I created a function to check and return the flag status by calling the manager within said function. Here I was able to await the featureflagmanager, of course.

3. I created a hook called, 'checkFeatureFlagAfterRead to be called at the collection level. (I added the name of the feature flag too of course, but obfuscating here). It was crucial that I returned this structure in this hook, adding the flag status along with the docs:

` return {
        ...doc,
        featureFlags: {
            myFeatureFlag: isEnabled
        }
    };`
* isEnabled is just a boolean value.

The afterRead hook fires off once the data is loaded on the page, but before the page renders (to my understanding)."

4. Then, in my Tab conditional statement: 
`  admin: {
    condition: (data) => {
      return !data?.featureFlags?.myFeatureFlag //Falsey logic for our situation
    },
  },`
  
  This is what ended up working 💯 I did my best explaining it, maybe it will help someone in the future haha!
